### PR TITLE
Travis and Coveralls added

### PR DESCRIPTION
**Pull request by PunyW**

### CODE REVIEW

**Comment by jamox** [comment]
Might be worth to rebase attempts to get Travis working into single commit.
And maybe do same for pit and cobertura.

**Comment by jamox** [comment]
Travis badge points to incorrect repo, should it be merged.
Wonder whether there is a nice solution to show different badges for different repos...

File: pom.xml [hunk 1]
```diff
@@ -6,12 +6,76 @@
...
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <version>2.3</version>
+		  <executions>
+		     <!-- Run shade goal on package phase -->
+		    <execution>
+			<phase>package</phase>
+			<goals>
+				<goal>shade</goal>
```
Comment (mpartel) [inline pom.xml]: tabs -> spaces and autoformat the entire file please. Please make your editor only produce spaces.

File: tmc-langs-ant/pom.xml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            <artifactId>junit</artifactId>
+            <version>4.10</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
+            <version>2.4</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (mpartel) [inline tmc-langs-ant/pom.xml]: Is there anything you need from this that isn't present in Guava? I didn't notice this used anywhere.

File: tmc-langs-ant/src/main/java/fi/helsinki/cs/tmc/langs/ant/AntPlugin.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.logging.Logger;
+
+/**
+ *
+ * @author TMC-langs
+ */
```
Comment (mpartel) [inline tmc-langs-ant/src/main/java/fi/helsinki/cs/tmc/langs/ant/AntPlugin.java]: remove

File: tmc-langs-ant/src/test/java/fi/helsinki/cs/tmc/langs/ant/AntPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static org.junit.Assert.*;
+
+/**
+ *
+ * @author alpa
+ */
```
Comment (mpartel) [inline tmc-langs-ant/src/test/java/fi/helsinki/cs/tmc/langs/ant/AntPluginTest.java]: remove

File: tmc-langs-ant/src/test/java/fi/helsinki/cs/tmc/langs/ant/AntPluginTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Test
+    public void returnsFalseForExercisesWithoutBuildFile() {
+        assertFalse(antPlugin.isExerciseTypeCorrect(getPath("non_ant_project")));
+    }
+
+    private Path getPath(String location) {
+        Path path = null;
+        try {
+             path = Paths.get(getClass().getResource(File.separatorChar + location).toURI());
```
Comment (mpartel) [inline tmc-langs-ant/src/test/java/fi/helsinki/cs/tmc/langs/ant/AntPluginTest.java]: indentation

File: tmc-langs-ant/src/test/java/fi/helsinki/cs/tmc/langs/ant/AntPluginTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        } catch (URISyntaxException e) {
+            e.printStackTrace();
```
Comment (mpartel) [inline tmc-langs-ant/src/test/java/fi/helsinki/cs/tmc/langs/ant/AntPluginTest.java]: Rethrow as runtime exception. With Guava, it's easy:
`throw Throwables.propagate(e);` ([see here](http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/base/Throwables.html#propagate(java.lang.Throwable)))

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java [hunk 7]
```diff
@@ -50,10 +51,10 @@
      *
      * <p>
      * Must return null if the given path is not a valid exercise directory
-     * for this langauge.
+     * for this language.
      *
      * @param path The path of the exercise directory.
-     * @param exerciseName This must be set as the returned value's `name`.
+     * @param exerciseName This must be set as the name of the returned value.
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java]: let's say "returned exercise"

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java [hunk 8]
```diff
@@ -70,7 +71,7 @@
      * Prepares a submission for processing in the sandbox.
      *
      * <p>
-     * The destination path is initialized with the original exercise as it
+     * The destination path is initialised with the original exercise as it
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java]: How British of you :)

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java [hunk 9]
```diff
@@ -105,8 +106,16 @@
...
      * The solution usually has stubs and special comments stripped.
      *
-     * @param path A path to a directory where the original exericse has been
+     * @param path A path to a directory where the original exercise has been
      * copied. This method should modify the contents of this directory.
      */
     public void prepareSolution(Path path);
+
+    /**
+     * Check if the exercises project type corresponds with the language plugin type.
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java]: "exercise's"

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java [hunk 10]
```diff
@@ -105,8 +106,16 @@
...
+     *
+     * @param path The path to the exercise directory.
+     * @return True if language plugin
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java]: ?

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java [hunk 11]
```diff
@@ -105,8 +106,16 @@
...
+     */
+    public boolean isExerciseTypeCorrect(Path path);
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java]: Doesn't `scanExercise` already do this?

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package fi.helsinki.cs.tmc.langs;
+
+import java.nio.file.Path;
+
+/**
+ *
+ * @author TMC-langs
+ */
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java]: remove

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+public abstract class LanguagePluginAbstract implements LanguagePlugin {
+
+    protected final String languageName;
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java]: Seems unnecessary. I see no benefit over having the subclass just override getLanguageName() itself.

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        this.languageName = languageName;
+    }
+           
+    @Override
+    public String getLanguageName() {
+       return languageName;
+    }
+    
+    @Override
+        public void prepareSubmission(Path submissionPath, Path destPath) {
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java]: Indentation. Same below.

File: tmc-langs-framework/src/test/resources/arith_funcs/src/Arith.java~ [hunk 15]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (mpartel) [inline tmc-langs-framework/src/test/resources/arith_funcs/src/Arith.java~]: Why does the file name end in a tilde?

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/ProjectType.java [hunk 16]
```diff
@@ -0,0 +1,21 @@
+package fi.helsinki.cs.tmc.langs.util;
+
+import fi.helsinki.cs.tmc.langs.LanguagePlugin;
+import fi.helsinki.cs.tmc.langs.ant.AntPlugin;
+
+/**
+ * All the possible project types
+ */
+public enum ProjectType {
```
Comment (mpartel) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/ProjectType.java]: Ideally I'd use Guice multibindings or something equivalent here to get a true plugin architecture. Enums are inflexible.

It's OK to do this later if you want.

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/ProjectTypeHandler.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+     *
+     * <p>
+     * Iterate through all language plugins to find one that recognises the
+     * project as their type.
+     *
+     * @param path The path to the exercise directory.
+     * @return The project type, or null if none.
+     */
+    public ProjectType getProjectType(Path path) {
+        for(ProjectType type : projectTypes.keySet()) {
```
Comment (mpartel) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/ProjectTypeHandler.java]: The hashmap seems unnecessary. You could just iterate over ProjectType.values() and get the plugin from the ProjectType.

Also, our coding style says there should be a space after keywords like `for` and `if` (but not after method names).

File: tmc-langs-util/src/test/java/fi/helsinki/cs/tmc/langs/util/ProjectTypeTest.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+
+import fi.helsinki.cs.tmc.langs.ant.AntPlugin;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class ProjectTypeTest {
+
+    @Test
+    public void testProjectTypeJAVA_ANT() {
```
Comment (mpartel) [inline tmc-langs-util/src/test/java/fi/helsinki/cs/tmc/langs/util/ProjectTypeTest.java]: I question the necessity of this.

**Comment by mpartel** [comment]
Also please rebase to clean up the commit history a bit.

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/ProjectType.java (hunk 16 above)
Comment (PunyW) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/ProjectType.java]: I'll take a look at Guice and try to implement it during this sprint if we have time. 

Thanks for the advice!

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java [hunk 19]
```diff
@@ -79,28 +106,16 @@
...
-     *   <li>Place any temporary files in the /tmp/build directory.
-     *       For instance, downloads should go there. These files are excluded
-     *       from the sandbox image.
-     *   <li>Do not depend on the execution order of language plugins.
-     * </ul>
+     * @param path The path to the exercise directory.
+     * @return True if given path is valid directory for this language  plugin
      */
-    public void prepareSandboxChroot();
+    public boolean isExerciseTypeCorrect(Path path);
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePlugin.java]: Isn't this a subset of what scanExercise does?

File: tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+package fi.helsinki.cs.tmc.langs;
+
+import java.nio.file.Path;
+
+public abstract class LanguagePluginAbstract implements LanguagePlugin {
+
+    @Override
+    public String getLanguageName() {
+        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+    }
```
Comment (mpartel) [inline tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java]: This probably won't get implemented here so leave abstract?

**Comment by mpartel** [comment]
I think this needs to be rebased on top of tmc-langs/master to be mergable. See `gitk --all`.

The commit history ended up a bit messy but I know it's quite troublesome to fix up after the fact so let's not spend any more time on it. In the future, try to merge small fixup commits earlier. It's much easier that way. But don't worry about it *too* much. And as always, make a backup before using rebase unless you're really confident :)

**Pull request merged**